### PR TITLE
[win32] ToolBar gets resized when mixing ToolItems with text and icon eclipse-platform/eclipse.platform.ui#443

### PR DESCRIPTION
**Pull request by Christopher-Hermann**
Fix for issue eclipse-platform/eclipse.platform.ui#443

### CODE REVIEW

**Comment by Christopher-Hermann** [comment]
Test is also failing in master, not related to the changes in this PR.
https://ci.eclipse.org/platform/job/eclipse.platform.swt/job/master/726/

**Comment by vogella** [comment]
@sratz can you review?

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/ToolBar.java [hunk 1]
```diff
@@ -735,6 +735,46 @@ void layoutItems () {
...
+	/*
+	 * Feature in Windows. When a tool bar has style SWT.FLAT, SWT.HORIZONTAL and
+	 * contains both text and images, Window leaves too much padding around the button.
+	 * This affects every button in the tool bar and makes the preferred height too big.
+	 * The fix is to set the TBSTYLE_LIST when the tool bar contains both text and images.
+	 */
+	if (OS.IsAppThemed()) {
+		if ((style & SWT.FLAT) != 0 && (style & SWT.HORIZONTAL) != 0) {
+			boolean hasText = false, hasImage = false;
+			for (ToolItem item : items) {
```
Comment (sratz) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/ToolBar.java:748]: if a SINGLE toolbar item has both image + text we do not want to do this condensing.

**Review by sratz — CHANGES_REQUESTED** [review]

**Comment by marcushoepfner** [comment]
Regarding the failing test Test_org_eclipse_swt_custom_StyledText.test_caretSizeAndPositionVariableGlyphMetrics

It is this assert:
`assertEquals(text.getLineHeight(0) - text.getCaret().getSize().y, text.getCaret().getBounds().y);`

On my Mac locally this is:
text.getLineHeight(0):103
text.getCaret().getSize().y:16
text.getCaret().getBounds().y:87

The jenkins is a linux where the values are:
text.getLineHeight(0):104
text.getCaret().getSize().y:19
text.getCaret().getBounds().y:170

As you can see here: https://ci.eclipse.org/platform/job/eclipse.platform.swt/job/PR-517/lastCompletedBuild/testReport/org.eclipse.swt.tests.junit/Test_org_eclipse_swt_custom_StyledText/test_caretSizeAndPositionVariableGlyphMetrics/

Is the assert wrong or is the productive code having issues on Linux?

**Comment by marcushoepfner** [comment]
> Regarding the failing test Test_org_eclipse_swt_custom_StyledText.test_caretSizeAndPositionVariableGlyphMetrics
> 
> It is this assert: `assertEquals(text.getLineHeight(0) - text.getCaret().getSize().y, text.getCaret().getBounds().y);`
> 
> On my Mac locally this is: text.getLineHeight(0):103 text.getCaret().getSize().y:16 text.getCaret().getBounds().y:87
> 
> The jenkins is a linux where the values are: text.getLineHeight(0):104 text.getCaret().getSize().y:19 text.getCaret().getBounds().y:170
> 
> As you can see here: https://ci.eclipse.org/platform/job/eclipse.platform.swt/job/PR-517/lastCompletedBuild/testReport/org.eclipse.swt.tests.junit/Test_org_eclipse_swt_custom_StyledText/test_caretSizeAndPositionVariableGlyphMetrics/
> 
> Is the assert wrong or is the productive code having issues on Linux?

The bounds of the caret are set via:
`Thread [main] (Suspended (breakpoint at line 323 in Caret))	
	Caret.setBounds(int, int, int, int) line: 323	
	StyledText.setCaretLocations(Point[], int) line: 8524	
	StyledText.setCaretLocations() line: 8439	
	StyledText.setStyleRanges(int, int, int[], StyleRange[], boolean) line: 10230	
	StyledText.setStyleRange(StyleRange) line: 10038	
	Test_org_eclipse_swt_custom_StyledText.test_caretSizeAndPositionVariableGlyphMetrics() line: 4878	
  [...]`
where as Caret.setBounds(int, int, int, int) is called with 2, 87, 0, 16

@mickaelistria is that something you could kindly check?

**Comment by akurtakov** [comment]
Failing test is linux specific and known failure so please ignore for this patch.

**Review by sratz — APPROVED** [review]
The precondition should be specific enough to not have any side-effects in other situations.

So this looks good to me.

**Pull request merged**